Improvements and refactoring of Nucleotide.java

vruano opened the pull request:
Change method names `valueOf` and `toBase` for `decode` and `encodeAsByte`. Reason is that `valueOf` (taking an string for example) has a bit of a meaning in Java Enums (e.g. it should result in a NotSuchElementException if such a constant does not exist and we don't do that here (returns Nucleotide#X).

Apart from that I added a few more methods to cover for functionality that I need as part of a larger pull-request for SV.

vruano commented:
@samuelklee please take a look.

vruano commented:
There were travis issue solvable by rebasing (automatic merge didn't compile).

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -14,19 +14,106 @@
+
+    // Extended codes:
+    // CODE(included nucs)
+    R(A,G), // Purine
```
Nitpicking, but perhaps make the descriptions here a little more consistent in style?  Currently, some are plural, some end in periods, etc.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -56,30 +145,198 @@ public byte toBase() {
+    }
+
+    /**
+     * Checks whether the nucleotide refer to an ambiguous base.
```
refer -> refers

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -56,30 +145,198 @@ public byte toBase() {
+    /**
+     * Checks whether this nucleotide code encloses all possible nucleotides for another code.
+     * @param other the other nucleotide to compare to.
+     * @return {@code true} iff any nucleotide in {@code other} is enclosed it this code.
```
it -> in

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -56,30 +145,198 @@ public byte toBase() {
+    }
+
+    /**
+     * Checks whether to base encodings make reference to the same {@link #Nucleotide}
```
to -> two

samuelklee replied:
Still need to fix this.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -56,30 +145,198 @@ public byte toBase() {
+     *     each possible nucleotide in this code.
+     * </p>
+     * <p>
+     *     The complement of the {@link #INVALID} nucleotide its itself.
```
its -> is

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -56,30 +145,198 @@ public byte toBase() {
+    }
+
+    /**
+     * Returns the instance that would include all possible transition mutation from this one.
```
mutation -> mutations

samuelklee replied:
(or perhaps: transition mutation -> transitions)

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -56,30 +145,198 @@ public byte toBase() {
+    }
+
+    /**
+     * Returns the instance that would include all possible tranversions mutation from this one.
```
similar here, transversions mutation -> transversion mutations (or transversions)

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -131,8 +388,14 @@ public void clear() {
             return LongStream.of(counts).sum();
         }
     }
+
```
white space

samuelklee commented on src/test/java/org/broadinstitute/hellbender/utils/NucleotideUnitTest.java:
```diff
@@ -19,65 +19,278 @@
-        Assert.assertEquals(Nucleotide.X.toBase(), (byte)'X');
+    @Test(dataProvider = "values")
+    public void testEncodeAsByte(final Nucleotide nuc) {
+        // Will always use the first letter of the constant as the one byt
```
one byt -> ?

samuelklee commented on src/test/java/org/broadinstitute/hellbender/utils/NucleotideUnitTest.java:
```diff
@@ -19,65 +19,278 @@
             }
-            Assert.assertSame(Nucleotide.valueOf(i), expected, "Failed with base " + i + " returning nucleotide " + Nucleotide.valueOf(i));
+        }
+
```
white space

magicDGS commented:
Hello @vruano - the code that you have here is quite nice, and I would like to include some bits on the next version of htsjdk ([htsjdk-next-beta](https://github.com/samtools/htsjdk-next-beta)). Do you have any problem with that?

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -64,58 +65,87 @@
+            final int lowerCaseIndex = nucleotide.lowerCaseByteEncoding & 0xFF;
+            final int upperCaseIndex = nucleotide.upperCaseByteEncoding & 0xFF;
+            maskToValue[nucleotide.mask]
+                    = baseToValue[lowerCaseIndex]
```
Not sure if this adheres to Java style guide.

vruano replied:
I quick google does not show anything against multiple assigns, perhaps you can post a link to it.

I will break into two lines the assignations to maskToValue from the other two to baseToValue as they go into different arrays, is perhaps more clear that way.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -64,58 +65,87 @@
     private final int mask;
     private final boolean isStandard;
+
+    // Some properties initialized after construction as these depends on some static arrays.
```
depends -> depend

vruano replied:
Ok.

samuelklee commented:
A couple more minor comments.  I will take your word that your changes improve performance!

vruano commented:
@samuelklee 
I addressed your comments + improved docs + added support for char and String encodings. Won't add anything else after your final review, promissed : P

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -3,22 +3,49 @@
  * <p>
  *    This enumeration not only contains standard (non-ambiguous) nucleotides, but also
- *    values to represent ambiguous and invalid codes.
+ *    values to represent ambiguous and an invalid nucleotide call ({@link #X} aka {@link #INVALID}.
```
Perhaps "...but also contains ambiguous nucleotides, as well as a code {@link #X} (a.k.a. {@link #INVALID}) for invalid nucleotide calls."

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -3,22 +3,49 @@
  *
+ * <p>
+ *     You can query whether a value refers to a non-ambiguous nucleotide with {@link #isStandard()} or
+ *     {@link #isAmbiguous()} depending of your preference. Notice that the special value {@link #X}
```
`X` is grouped together with the ambiguous nucleotides below in line 95.  Perhaps add a line break to avoid confusion.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -3,22 +3,49 @@
+ * <p>
+ *     Querying the {@link #X} value for its {@link #complement}, {@link #transition} or
+ *     {@link #transversion} or using it in other operations
+ *     likes {@link #intersect} would result in returning also a {@link #X}; similar to {@link Double#NaN} in
```
likes intersect would result in returning also a X -> like (or perhaps "such as") intersect will return an X

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -3,22 +3,49 @@
+ *     Finally, notice that there is no code of the "gap nucleotide" that may appear in aligned sequences as in fact that is not a nucleotide.
+ *     A base encoding using the typical gap representation such as '.' or '-' would
+ *     be interpreted as an {@link #INVALID} (i.e. {@link #X}) call which is probably not what you want.
+ *     So code that mus support those will need to do so outside this {@code enum}.
```
So code that mus support - > So code to support?

samuelklee replied:
BTW, thanks for adding all of this documentation!

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -3,22 +3,49 @@
+ *     convenient longer form names constant aliases (e.g. {@link #ADENINE} for {@link #A}, {@link #PURINE} for {@link #R}, etc.).
+ * </p>
+ * <p>
+ *     Finally, notice that there is no code of the "gap nucleotide" that may appear in aligned sequences as in fact that is not a nucleotide.
```
code of -> code for

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -89,26 +128,25 @@
         Arrays.fill(baseToValue, INVALID);
         Arrays.fill(baseToOrdinal, INVALID.ordinal());
         for (final Nucleotide nucleotide : values) {
+            // Notice that {@code "x & 0xFF"} is needed instead of {@code "(int)x" as
```
Actually, can you just use `Byte.toUnsignedInt` everywhere?

It may be worth checking if there are human-readable methods for the other bitwise operations you use throughout, or extracting your own if necessary.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -120,17 +158,36 @@
+
+    /**
+     * Holds the lower-case {@code char} representation for this nucleotide.
+     * This is typically the lower-case version of the enum constant name only character.
```
name only character -> name, as a character?

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -120,17 +158,36 @@
 
+    /**
+     * Holds the upper-case {@code char} representation for this nucleotide.
+     * This is typically the upper-case version of the enum constant name only character.
```
same here

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -151,27 +212,56 @@
+
+    /**
+     * Returns this nucleotide's exclusive upper-case {@code byte} encoding.
+     * @return <i>ditto</i>.
```
I think you can probably just leave out the `@return` lines here and below.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -184,12 +274,37 @@ public static Nucleotide decode(final byte base) {
-    public static Nucleotide decode(final char base) {
-        return baseToValue[base & 0xFF];
+    public static Nucleotide decode(final char ch) {
+        if ((ch & 0xFF00) != 0) {
```
Perhaps extract the operations used in this method.

samuelklee commented on src/main/java/org/broadinstitute/hellbender/utils/Nucleotide.java:
```diff
@@ -184,12 +274,37 @@ public static Nucleotide decode(final byte base) {
+     * Transform a single-letter character string into the corresponding nucleotide.
+     * <p>
+     *    {@code Null}, empty or multi-letter input will result in an {@link IllegalArgumentException}.
+     *    These are not simply invalid encodings as the fact that are not a single character is
```
encodings as the fact that are not a single character  -> encodings, and the fact that they are not a single character?

samuelklee commented on src/test/java/org/broadinstitute/hellbender/utils/NucleotideUnitTest.java:
```diff
@@ -31,7 +31,70 @@ public void testEncodeAsByte(final Nucleotide nuc) {
+
+    /**
+     * Checks the assumption that each nuc canonical name is a single upper-case letter.
+     * If this test fails that would  indicates that you are modifying {@link Nucleotide} in a way
```
that would indicates -> that would indicate

samuelklee commented on src/test/java/org/broadinstitute/hellbender/utils/NucleotideUnitTest.java:
```diff
@@ -329,6 +426,56 @@ public void testAddingAllAtOnce(final byte[] bases) {
+            Assert.assertEquals(subject.get(n), (long) shadow.getOrDefault(n, 0));
+        }
+        Assert.assertEquals(subject.sum(), shadow.values().stream().mapToLong(l -> l).sum());
+    }
```
Thanks for adding these tests!
